engine/consensus/sealing/core.go

```diff
@@ -112,6 +113,104 @@ func NewCore(
+// Repopulating is split into two parts:
+// 1) traverse forward all finalized blocks starting from last sealed block till we reach last finalized block . (lastSealedHeight, lastFinalizedHeight]
+// 2) traverse forward all unfinalized(pending) blocks starting from last finalized block.
+// For each block that is being traversed we will collect execution results and process them using sealing.Core.
+func (c *Core) RepopulateAssignmentCollectorTree(payloads storage.Payloads) error {
+	finalizedSnapshot := c.state.Final()
+	finalized, err := finalizedSnapshot.Head()
+	if err != nil {
+		return fmt.Errorf("could not retrieve finalized block: %w", err)
+	}
+	finalizedID := finalized.ID()
+
+	// Get the latest sealed block on this fork, ie the highest block for which
+	// there is a seal in this fork.
+	latestSeal, err := c.seals.ByBlockID(finalizedID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve parent seal (%x): %w", finalizedID, err)
+	}
+
+	latestSealedBlockID := latestSeal.BlockID
+	latestSealedBlock, err := c.headers.ByBlockID(latestSealedBlockID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve latest sealed block (%x): %w", latestSealedBlockID, err)
+	}
+
+	// usually we start with empty collectors tree, prune it to minimum height
+	_, err = c.collectorTree.PruneUpToHeight(latestSealedBlock.Height)
+	if err != nil {
+		return fmt.Errorf("could not prune execution tree to height %d: %w", latestSealedBlock.Height, err)
+	}
+
+	blocksProcessed := uint64(0)
+	totalBlocks := finalized.Height - latestSealedBlock.Height
+
+	// resultProcessor adds _all known_ results for the given block to the assignment collector tree
+	resultProcessor := func(header *flow.Header) error {
+		blockID := header.ID()
+		payload, err := payloads.ByBlockID(blockID)
+		if err != nil {
+			return fmt.Errorf("could not retrieve index for block (%x): %w", blockID, err)
+		}
+
+		for _, result := range payload.Results {
+			// TODO: change this when migrating to sealing & verification phase 3.
+			// Incorporated result is created this way only for phase 2.
+			incorporatedResult := flow.NewIncorporatedResult(result.BlockID, result)
+			err = c.ProcessIncorporatedResult(incorporatedResult)
+			if err != nil {
+				return fmt.Errorf("could not process incorporated result for block %s: %w", blockID, err)
+			}
+		}
+
+		blocksProcessed++
+		if (blocksProcessed%20) == 0 || blocksProcessed >= totalBlocks {
+			c.log.Info().Msgf("%d/%d have been loaded to collector tree", blocksProcessed, totalBlocks)
+		}
+
+		return nil
+	}
+
+	c.log.Info().Msgf("there are %d finalized and unsealed blocks in total to reload into collector tree with assignment",
+		totalBlocks)
+
+	// traverse chain forward to collect all execution results that were incorporated in this fork
+	// starting from finalized block and finishing with latest sealed block
```

Comment:
since we are traversing forward, we actually start with processing the direct child of the last finalized block and end the processing with the last finalized block:
```suggestion
	// we start with processing the direct child of the last finalized block and end with the last finalized block
```